stream_batch.go

```diff
@@ -0,0 +1,220 @@
+		if err != nil {
+			return err
+		}
+		oldProgress[string(stream)] = progress.Time
+	}
+
+	minTime := oldProgress[string(sbr.streams[0])]
+	for _, progress := range oldProgress {
+		if compareTimeuuid(minTime, progress) > 0 {
+			minTime = progress
+		}
+	}
+	if (minTime != gocql.UUID{}) {
+		sbr.lastTimestamp = minTime
+	}
+
+	sbr.config.Logger.Printf("running batch starting from %v", sbr.lastTimestamp.Time())
+
+	input := CreateChangeConsumerInput{
+		TableName: baseTableName,
+		StreamIDs: sbr.streams,
+	}
+	consumer, err := sbr.config.ChangeConsumerFactory.CreateChangeConsumer(input)
+	if err != nil {
+		sbr.config.Logger.Printf("error while creating change consumer (will quit): %s", err)
+	} else {
+		defer consumer.End()
+	}
+
+	crq := newChangeRowQuerier(sbr.config.Session, sbr.streams, sbr.keyspaceName, sbr.tableName)
+
+	// sbr.config.Logger.Printf("starting stream processor loop for %v", sbr.streams)
+outer:
+	for {
+		timeWindowEnd := sbr.lastTimestamp.Time().Add(sbr.config.Advanced.QueryTimeWindowSize)
+		confidenceWindowEnd := time.Now().Add(-sbr.config.Advanced.ConfidenceWindowSize)
+
+		if timeWindowEnd.After(confidenceWindowEnd) {
+			timeWindowEnd = confidenceWindowEnd
+		}
+
+		pollEnd := gocql.MinTimeUUID(timeWindowEnd)
+
+		var (
+			err      error
+			rowCount int
+		)
+
+		readUpTo := pollEnd
+
+		if compareTimeuuid(sbr.lastTimestamp, pollEnd) < 0 {
+			// Set the time interval from which we need to return data
+			var iter *changeRowIterator
+			iter, err = crq.queryRange(sbr.lastTimestamp, pollEnd)
+			if err != nil {
+				sbr.config.Logger.Printf("error while sending a query (will retry): %s", err)
+			} else {
+				var change Change
+				for {
+					streamCols, c := iter.Next()
+					if c == nil {
+						break
+					}
+
+					if c.GetOperation() == PreImage {
+						change.Preimage = append(change.Preimage, c)
+					} else if c.GetOperation() == PostImage {
+						change.Postimage = append(change.Postimage, c)
+					} else {
+						change.Delta = append(change.Delta, c)
+					}
+
+					if c.cdcCols.endOfBatch {
+						// Prevent changes older than the save point from being replayed again
+						if compareTimeuuid(oldProgress[string(change.StreamID)], streamCols.time) < 0 {
+							change.StreamID = streamCols.streamID
+							change.Time = streamCols.time
+							if err := consumer.Consume(change); err != nil {
+								// TODO: Does that make sense?
+								sbr.config.Logger.Printf("error while processing change (will quit): %s", err)
+								return err
+							}
+						} else {
+							sbr.config.Logger.Printf("skipping change due to it being too old (%v <= %v)",
+								streamCols.streamID, oldProgress[string(change.StreamID)].Time())
+						}
+
+						change.Preimage = nil
+						change.Delta = nil
+						change.Postimage = nil
+
+						// Update the last timestamp only after we processed whole batch
+						if compareTimeuuid(sbr.lastTimestamp, streamCols.time) < 0 {
+							sbr.lastTimestamp = streamCols.time
```

Comment:
I don't get this, don't we update `lastTimestamp` below?
```
 		sbr.lastTimestamp = readUpTo 
```

isn't this enough?

Reply:
Yes, we can get rid of it. Thanks for pointing out.